library/std/src/ffi/c_str.rs

```diff
@@ -1299,7 +1319,13 @@ impl CStr {
     #[stable(feature = "cstr_from_bytes", since = "1.10.0")]
     #[rustc_const_unstable(feature = "const_cstr_unchecked", issue = "none")]
     pub const unsafe fn from_bytes_with_nul_unchecked(bytes: &[u8]) -> &CStr {
-        &*(bytes as *const [u8] as *const CStr)
+        // SAFETY: `bytes` is a reference and so safe to cast to a pointer.
+        // Casting to CStr is safe because its internal represention is a [u8]
```

Comment:
```suggestion
        // Casting to CStr is safe because its internal representation is a [u8]
```

We should also note that the cast is safe only in std, where we have knowledge that CStr's layout is compatible with [u8] despite the lack of `#[repr(transparent)]`.